internal/network/connectivity_groups.go

```diff
@@ -234,53 +255,170 @@ func createHostGroupCandidate(hostIndex, numHosts int, cMap connectivityMap) gro
 	}
 }
 
-/*
- * Create connectivity map from host list.  It is the information if a host has connectivity to other host on specific
- * CIDR (network)
- */
-func createMachineCidrConnectivityMap(cidr string, hosts []*models.Host, idToIndex map[strfmt.UUID]int) (connectivityMap, error) {
-	ret := make(connectivityMap)
+type hostQueryFactory interface {
+	create(h *models.Host) (hostQuery, error)
+}
+
+type hostQuery interface {
+	next() strfmt.UUID
+}
+
+type l2Query struct {
+	parsedCidr         *net.IPNet
+	current            int
+	connectivityReport models.ConnectivityReport
+}
+
+func (l *l2Query) next() strfmt.UUID {
+	for l.current != len(l.connectivityReport.RemoteHosts) {
+		rh := l.connectivityReport.RemoteHosts[l.current]
+		l.current++
+		for _, l2 := range rh.L2Connectivity {
+			ip := net.ParseIP(l2.RemoteIPAddress)
+			if ip != nil && l.parsedCidr.Contains(ip) && l2.Successful {
+				return rh.HostID
+			}
+		}
+	}
+	return ""
+}
+
+type l2QueryFactory struct {
+	parsedCidr *net.IPNet
+}
+
+func (l *l2QueryFactory) create(h *models.Host) (hostQuery, error) {
+	ret := l2Query{
+		parsedCidr: l.parsedCidr,
+	}
+	err := json.Unmarshal([]byte(h.Connectivity), &ret.connectivityReport)
+	if err != nil {
+		return nil, err
+	}
+	return &ret, nil
+}
+
+func newL2QueryFactory(cidr string) (hostQueryFactory, error) {
 	_, parsedCidr, err := net.ParseCIDR(cidr)
 	if err != nil {
 		return nil, err
 	}
-	for fromIndex, h := range hosts {
-		if h.Connectivity == "" {
+	return &l2QueryFactory{
+		parsedCidr: parsedCidr,
+	}, nil
+}
+
+type l3Query struct {
+	current            int
+	connectivityReport models.ConnectivityReport
+	nodesAddresses     map[strfmt.UUID]map[string]bool
+}
+
+func (l *l3Query) next() strfmt.UUID {
+	for l.current != len(l.connectivityReport.RemoteHosts) {
+		rh := l.connectivityReport.RemoteHosts[l.current]
+		l.current++
+		addresses, ok := l.nodesAddresses[rh.HostID]
+		if !ok {
 			continue
 		}
-		var connectivityReport models.ConnectivityReport
-		err = json.Unmarshal([]byte(h.Connectivity), &connectivityReport)
+		foundAddresses := make(map[string]bool)
+		for _, l3 := range rh.L3Connectivity {
+			_, foundAddress := addresses[l3.RemoteIPAddress]
+			if foundAddress && l3.Successful {
+				foundAddresses[l3.RemoteIPAddress] = true
+			}
+		}
+		if len(addresses) == len(foundAddresses) {
+			return rh.HostID
+		}
+	}
+	return ""
+}
+
+type l3QueryFactory struct {
+	nodesAddresses map[strfmt.UUID]map[string]bool
+}
+
+func (l *l3QueryFactory) create(h *models.Host) (hostQuery, error) {
+	ret := l3Query{
+		nodesAddresses: l.nodesAddresses,
+	}
+	err := json.Unmarshal([]byte(h.Connectivity), &ret.connectivityReport)
+	if err != nil {
+		return nil, err
+	}
+	return &ret, nil
+}
+
+func newL3QueryFactory(hosts []*models.Host, family AddressFamily) (hostQueryFactory, error) {
+	nodesAddresses := make(map[strfmt.UUID]map[string]bool)
+	for _, h := range hosts {
+		if h.Inventory == "" {
+			continue
+		}
```

Comment:
Won't it cause skipping the check for a host that doesn't have an inventory? Instead of failing connectivity validation for that host.

Reply:
It will cause to fail connectivity check for the host.  The logic is positive.  A host is regarded disconnected unless it we indicate in the code specifically that it is connected.